mypyc/genops.py

```diff
@@ -701,12 +702,101 @@ def visit_assignment_stmt(self, stmt: AssignmentStmt) -> Value:
+            res = self.coerce(res, target.type, res.line)
+            self.add(Assign(target.register, res, res.line))
+       else:
+           assert False, "Unsupported assignment target"
+
+    def in_place_op_attr(self, target: AssignmentTarget, rvalue_reg: Value, op: str, line: int) -> None:
+        method_name = self.translate_in_place_op(op)
+        assert method_name is not None, "can't find method for op %s" % op
+        target_value = self.read_from_target(target, line)
+        if isinstance(target_value.type, RInstance):
+            self.add(MethodCall(void_rtype,
+                                target_value,
+                                method_name,
+                                [rvalue_reg],
+                                line))
+        elif isinstance(target_value.type, RPrimitive):
+            res = self.binary_op(target_value, rvalue_reg, op + '=', line)
+            if isinstance(target.obj_type, RInstance):
+                res = self.coerce(res, target.type, line)
+                self.add(SetAttr(target.obj, target.attr, res, res.line))
+            else:
+                key = self.load_static_unicode(target.attr)
+                boxed_reg = self.box(res)
+                self.add(PrimitiveOp([target.obj, key, boxed_reg], py_setattr_op, res.line))
+        else:
+            assert False, "Unsupported assignment target"
+
+    def in_place_op_index(self, target: AssignmentTarget, rvalue_reg: Value, op: str, line: int) -> None:
+        method_name = self.translate_in_place_op(op)
+        assert method_name is not None, "can't find method for op %s" % op
+        target_value = self.read_from_target(target, line)
+        if isinstance(target_value.type, RInstance):
+            self.add(MethodCall(void_rtype,
+                                target_value,
+                                method_name,
+                                [rvalue_reg],
+                                line))
+        elif isinstance(target_value.type, RPrimitive):
+            res = self.binary_op(target_value, rvalue_reg, op + '=', line)
+            target_reg2 = self.translate_special_method_call(
+                   target.base,
+                   '__setitem__',
+                   [target.index, res],
+                   None,
+                   res.line)
+            assert target_reg2 is not None, target.base.type
+        else:
+            assert False, "Unsupported assignment target"
+
+    def in_place_op(self, target: AssignmentTarget, rvalue_reg: Value, op: str, line: int) -> None:
+       if isinstance(target, AssignmentTargetRegister):
+           self.in_place_op_register(target, rvalue_reg, op, line)
+       elif isinstance(target, AssignmentTargetAttr):
+           self.in_place_op_attr(target, rvalue_reg, op, line)
+       elif isinstance(target, AssignmentTargetIndex):
+           self.in_place_op_index(target, rvalue_reg, op, line)
+       else:
+           assert False, 'Unsupported assignment target'
```

Comment:
We shoudn't need to have separate cases for all of the different possible left hand sides. We should keep using `read_from_target` and `assign_to_target` to abstract all that away on both the reading from and assigning to side.